src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,130 @@
+import * as React from "react";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import { PASSWORD_KEY } from "../constants";
+
+type AppModeTypes = "normal" | "kill";
+type AuthStatusTypes = "pending" | "authenticated" | "notRequired";
+
+type AuthState = {
+  passcode: string | null;
+  appMode: AppModeTypes;
+  killModeEnabled: boolean;
+  authStatus: AuthStatusTypes;
+};
+
+type KillModeActions =
+  | { type: "setPasscode"; newPasscode: string | null }
+  | { type: "toggleAppMode"; newAppMode?: AppModeTypes }
+  | { type: "toggleKillModeEnabled"; newKillModeValue?: boolean }
+  | { type: "setAuthStatus"; newAuthStatus: AuthStatusTypes };
+
+function killModeReducer(state: AuthState, action: KillModeActions): AuthState {
+  switch (action.type) {
+    /***
+     * if passcode is set to null we need to set the auth status to 'notRequired'
+     *
+     * If a passcode is set, we want to set the authStatus to 'authenticated'
+     * If the auth status is set to 'pending' they will be navigated to the auth screen, to type in their password.
+     */
+    case "setPasscode":
+      const passcode = action.newPasscode;
+      if (passcode === null || validPasscode(passcode)) {
+        AsyncStorage.setItem(PASSWORD_KEY, passcode || "");
+        if (!passcode) {
+          return { ...state, passcode, authStatus: "notRequired" };
+        }
+        return { ...state, passcode, authStatus: "authenticated" };
+      }
+      throw new Error("Invalid New Password");
+    /***
+     * Does not require a new value in the dispatcher payload, but a new value is optional in the payload
+     * If there is no new value, the value is toggled (as there are only 2 values)
+     */
+    case "toggleAppMode":
+      return {
+        ...state,
+        appMode: !!action.newAppMode
+          ? action.newAppMode
+          : state.appMode === "kill"
+          ? "normal"
+          : "kill",
+      };
+    //will throw an error if the user does NOT have a password AND if the user is trying to set killMode to true
+    //With the killMode dispatch code, the user does not have to explicity set a value for kill mode
+    //If there is no value set, the kill mode just toggles, hence why I am checking if the value is
+    //explicitly being set to true, OR if it is not being set and kill mode is CURRENTLY false (and therfore will be set to true)
+    case "toggleKillModeEnabled":
+      if (
+        !state.passcode &&
+        (action.newKillModeValue ||
+          (action.newKillModeValue === undefined && !state.killModeEnabled))
+      ) {
+        throw new Error('Cannot enable "killMode" until password has been set');
+      }
+
+      return {
+        ...state,
+        killModeEnabled: !!action.newKillModeValue
+          ? action.newKillModeValue
+          : state.killModeEnabled
+          ? false
+          : true,
+      };
```

Comment:
refer to other [comment](https://github.com/digidem/mapeo-mobile/pull/923/files#r810353052) for suggestion on how to simplify this